Add support to custom log level through command line flag and environment variable

vitorhugoro1 opened the pull request:
- Added on `InstrumentationOption` the option to define log level.
- Implemented a custom severity enum, called level.go, with options to set log levels based on text using `zap.Level` as reference.
- Added on CLI the flag `-log-level` that accept one of `LogLevel` value, which can be see on [level.go](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/91e4e34ef58b7a05ca1e72110b6477f5cf942cbe/level.go) ex. `otel-go-instrumentation -log-level error`, the default value is `info`.
- Added support to `OTEL_LOG_LEVEL` which implements `LogLevel` to set visibility level through environment variable.
-  Changed multiples logs level to Debug to improve setup when using in production environment.  

- Closes #691

grcevski commented on instrumentation.go:
```diff
@@ -463,3 +474,12 @@ func WithResourceAttributes(attrs ...attribute.KeyValue) InstrumentationOption {
+// WithLogLevel returns an [InstrumentationOption] that will configure
+// an [Instrumentation] with the logger level visibility defined as inputed.
+func WithLogLevel(level string) InstrumentationOption {
+	return fnOpt(func(ctx context.Context, c instConfig) (instConfig, error) {
```
What do you think about parsing the log level here and erroring out if the user has specified something invalid? It would be nice to know for example if I've miss-spelled "debug", instead of silently switching to "info".

RonFed replied:
It looks like [ParseAtomicLevel](https://pkg.go.dev/go.uber.org/zap#ParseAtomicLevel) does some validation if we choose to use it (currently it looks like it is used in `newLogger`)

vitorhugoro1 replied:
Added the validation, what do you think?

grcevski replied:
Awesome!

grcevski commented on instrumentation_test.go:
```diff
@@ -172,6 +172,14 @@ func TestWithResourceAttributes(t *testing.T) {
 }
 
+func TestWithLogLevel(t *testing.T) {
+	c, err := newInstConfig(context.Background(), []InstrumentationOption{WithLogLevel("error")})
```
Thanks for adding these tests! Do you mind extending the testing to cover what happens if invalid options are specified or the configuration doesn't contain logging level?

This is related to my earlier comment that parsing of the log level should be done here, so that invalid command line input is detected.

vitorhugoro1 replied:
Created the test to cover the validation of a wrong logging level

MrAlias commented on instrumentation.go:
```diff
@@ -463,3 +474,12 @@ func WithResourceAttributes(attrs ...attribute.KeyValue) InstrumentationOption {
+
+// WithLogLevel returns an [InstrumentationOption] that will configure
+// an [Instrumentation] with the logger level visibility defined as inputed.
+func WithLogLevel(level string) InstrumentationOption {
```
Instead of accepting a string with special values, we should introduce a `LogLevel` type with `enum` values. We might want to use the [OTel levels](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-severitynumber). For example:

https://github.com/open-telemetry/opentelemetry-go/blob/main/log/severity.go

vitorhugoro1 replied:
Do you think we need to create something which uses numbers as reference like the one you linked?

MrAlias replied:
That seems reasonable to me.

MrAlias replied:
We already do something like this for offsetgen:

https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/e85ddd9de6c5829d6736d7ccefbc3733f2cb8ed8/internal/tools/inspect/cmd/offsetgen/main.go#L59

We could provided human readable flags for the CLI though (i.e. `--log-level info`).

vitorhugoro1 replied:
Yeah, I usually prefer something human readable to CLI because is more quick to identify.

vitorhugoro1 replied:
@MrAlias I have create a simple Level Enum for logs, what do you think?

MrAlias replied:
> @MrAlias I have create a simple Level Enum for logs, what do you think?

This still accepts a `string`.

```suggestion
func WithLogLevel(level Level) InstrumentationOption {
```

vitorhugoro1 replied:
@MrAlias  This case here expect an string because is the input when use flag or from env, so for properly work is these scenarios this need to be string

MrAlias replied:
The environment variable needs to be parsed before it is passed to this function then.

We need to use syntax here to ensure the best possible chance values passed will be valid. Using an arbitrary string with magic values is not appropriate.

vitorhugoro1 replied:
@MrAlias changed `WithLogLevel` to work on the behaviour you said

RonFed commented:
@vitorhugoro1 Do you plan on changing the level of the log line I mentioned in the issue?
We can use debug level in our tests in https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/e85ddd9de6c5829d6736d7ccefbc3733f2cb8ed8/.github/workflows/e2e/k8s/sample-job.yml#L29

pellared reviewed (commented):
Shouldn't we support setting via `OTEL_LOG_LEVEL` env var?

Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md#general-sdk-configuration

MrAlias commented:
> Shouldn't we support setting via `OTEL_LOG_LEVEL` env var?
> 
> Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md#general-sdk-configuration

Please see https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/691

That appears out of scope for this PR.

pellared commented:
> > Shouldn't we support setting via `OTEL_LOG_LEVEL` env var?
> > Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md#general-sdk-configuration
> 
> That is tracked in #691. It is out of scope for this PR.

This PR says

> Closes https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/691

vitorhugoro1 commented:
> > Shouldn't we support setting via `OTEL_LOG_LEVEL` env var?
> > Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md#general-sdk-configuration
> 
> That is tracked in #691. It is out of scope for this PR.

Yeah this PR will implement this change too, I take the approach just with flags because is the way I usually use log levels on container

vitorhugoro1 commented:
> @vitorhugoro1 Do you plan on changing the level of the log line I mentioned in the issue? In addition, we can use debug level in our tests at
> 
> https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/e85ddd9de6c5829d6736d7ccefbc3733f2cb8ed8/.github/workflows/e2e/k8s/sample-job.yml#L29

I changed some log level, not just in the line you mentioned but in another which is more suitable to debugging not exactly on a production setup.

Changed the sample job too to use debug level by default

vitorhugoro1 commented:
> Shouldn't we support setting via `OTEL_LOG_LEVEL` env var?
> 
> Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md#general-sdk-configuration

Added the supporting via environment with the name you suggest

grcevski commented on internal/pkg/instrumentation/manager.go:
```diff
@@ -199,23 +199,23 @@ func (m *Manager) load(target *process.TargetDetails) error {
 	// Load probes
 	for name, i := range m.probes {
-		m.logger.Info("loading probe", "name", name)
+		m.logger.V(-1).Info("loading probe", "name", name)
```
I think this might be useful to keep as info, to know which probes are activated.

vitorhugoro1 replied:
Changed to info

grcevski commented on internal/pkg/process/allocate.go:
```diff
@@ -81,7 +81,7 @@ func remoteAllocate(logger logr.Logger, pid int, mapSize uint64) (uint64, error)
 
 	defer func() {
-		logger.Info("Detaching from process", "pid", pid)
+		logger.V(-1).Info("Detaching from process", "pid", pid)
```
I'd keep this as info too.

vitorhugoro1 replied:
Changed to info

grcevski commented on internal/pkg/process/discover.go:
```diff
@@ -70,16 +70,16 @@ func (a *Analyzer) DiscoverProcessID(ctx context.Context, target *TargetArgs) (i
 			pid, err := a.findProcessID(target, proc)
 			if err == nil {
-				a.logger.Info("found process", "pid", pid)
+				a.logger.V(-1).Info("found process", "pid", pid)
```
Same here, keep this under Info.

vitorhugoro1 replied:
Changed to info

RonFed commented on cli/main.go:
```diff
@@ -68,8 +68,11 @@ func newLogger() logr.Logger {
+	var logLevel string
 
 	flag.BoolVar(&globalImpl, "global-impl", false, "Record telemetry from the OpenTelemetry default global implementation")
+	flag.StringVar(&logLevel, "log-level", "info", "Define log visibility level")
```
```suggestion
	flag.StringVar(&logLevel, "log-level", "", "Define log visibility level")
```

I think we should not set the default value here as info. Doing so will cause `OTEL_LOG_LEVEL=DEBUG` to be ignored since we add `WithLogLevel` after `WhitEnv` this is expected, but from a user who is running the agent image point of view this is not expected.
In addition we set the default level to info in `newInstConfig` if it is not define but env or by option.

vitorhugoro1 replied:
Yeah, makes sense what you say, changed to use like you recommended.

I have already set info as default on `newInstConfig`.

RonFed commented on cli/main.go:
```diff
@@ -98,6 +101,8 @@ func main() {
 		instOptions = append(instOptions, auto.WithGlobal())
 	}
 
+	instOptions = append(instOptions, auto.WithLogLevel(logLevel))
```
```suggestion
	if logLevel != "" {
		instOptions = append(instOptions, auto.WithLogLevel(logLevel))
	}
```

This relates to my other comment in this file

vitorhugoro1 replied:
Changed

vitorhugoro1 commented on instrumentation.go:
```diff
@@ -371,6 +385,11 @@ func WithEnv() InstrumentationOption {
 			c.serviceName = name
 			c.additionalResAttrs = append(c.additionalResAttrs, attrs...)
 		}
+		if l, ok := lookupEnv(envLogLevelKey); ok {
```
Updated

vitorhugoro1 commented:
@pellared @MrAlias @grcevski it's okay now for you guys?

grcevski reviewed (approved):
LGTM!

MrAlias commented on internal/pkg/log/level.go:
```diff
@@ -0,0 +1,104 @@
+	LevelWarn Level = "warn"
+	// Log Level [error].
+	LevelError Level = "error"
+	// Log Level [dpanic].
+	LevelDPanic Level = "dpanic"
```
What is dpanic?

vitorhugoro1 replied:
I will remove this cases

vitorhugoro1 replied:
I removed these options

MrAlias commented on internal/pkg/log/level.go:
```diff
@@ -0,0 +1,104 @@
+	// Log Level [panic].
+	LevelPanic Level = "panic"
+	// Log Level [fatal].
+	LevelFatal Level = "fatal"
```
What is the difference between panic and fatal? Shouldn't these both be terminations of the application?

vitorhugoro1 replied:
These two is to replicate what we see on zapcore.levels which is our base config

MrAlias replied:
I would replicate [OTel logging levels](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-severitynumber) if we are going to model this on a preexisting structure.

vitorhugoro1 replied:
I will remove these too

vitorhugoro1 replied:
I removed these options because they don't seem to make sense for the project

MrAlias commented on internal/pkg/log/level.go:
```diff
@@ -0,0 +1,104 @@
+	"fmt"
+)
+
+type Level string
```
Missing comment.

vitorhugoro1 replied:
Added comment

MrAlias commented on internal/pkg/log/level.go:
```diff
@@ -0,0 +1,104 @@
+type Level string
+
+const (
+	// Log Level [debug].
+	LevelDebug Level = "debug"
+	// Log Level [info].
+	LevelInfo Level = "info"
+	// Log Level [warn].
+	LevelWarn Level = "warn"
+	// Log Level [error].
+	LevelError Level = "error"
+	// Log Level [dpanic].
+	LevelDPanic Level = "dpanic"
+	// Log Level [panic].
+	LevelPanic Level = "panic"
+	// Log Level [fatal].
```
These comments do not explain the levels. Please update so they communicate the logging strata they intend to enable.

vitorhugoro1 replied:
Added comments to levels

MrAlias commented on internal/pkg/log/level.go:
```diff
@@ -0,0 +1,104 @@
+	"fmt"
+)
+
+type Level string
```
This needs to be in the exported `auto` package so users can reference it.

MrAlias replied:
This lives in an internal package currently. It needs to be moved to the user accessible `go.opentelemetry.io/auto` package instead.

vitorhugoro1 replied:
@MrAlias nice, changed to `auto` package the level

vitorhugoro1 commented:
@MrAlias any things needed here yet?

RonFed commented:
@vitorhugoro1 bump 😄

vitorhugoro1 commented on instrumentation.go:
```diff
@@ -463,3 +482,17 @@ func WithResourceAttributes(attrs ...attribute.KeyValue) InstrumentationOption {
+// an [Instrumentation] with the logger level visibility defined as inputed.
+func WithLogLevel(level Level) InstrumentationOption {
+	return fnOpt(func(ctx context.Context, c instConfig) (instConfig, error) {
+		if err := level.UnmarshalText([]byte(level.String())); err != nil {
```
Added a validate method to cover this case

vitorhugoro1 commented on instrumentation.go:
```diff
@@ -371,6 +385,11 @@ func WithEnv() InstrumentationOption {
 		}
+		if l, ok := lookupEnv(envLogLevelKey); ok {
+			var e error
+			c.logLevel, e = ParseLevel(l)
```
Changed

vitorhugoro1 commented on instrumentation.go:
```diff
@@ -204,6 +213,10 @@ func newInstConfig(ctx context.Context, opts []InstrumentationOption) (instConfi
 		c.sampler = trace.AlwaysSample()
 	}
 
+	if c.logLevel == "" {
```
Changed

vitorhugoro1 commented on level.go:
```diff
@@ -0,0 +1,87 @@
+
+const (
+	// Log Level [debug] Print all logs generate by instrumentation.
+	LevelDebug Level = "debug"
```
Applied the suggested

vitorhugoro1 commented on instrumentation.go:
```diff
@@ -68,8 +70,17 @@ type Instrumentation struct {
+func newLogger(logLevel Level) logr.Logger {
+	level, err := zap.ParseAtomicLevel(logLevel.String())
+	if err != nil {
+		level, _ = zap.ParseAtomicLevel(LevelInfo.String())
```
Added the log

MrAlias replied:
Is there a missing commit? This still stores the returned error as `_`.

vitorhugoro1 replied:
Oh this err is from zapcore itself, here I manually force the info. The validation for change to Info is make on the added line 98, where I catch if the first `ParseAtomicLevel` generated some error and show the message to user

MrAlias replied:
Understood. The issue is though that you will have a user that, say, sets the "trace" logging level expecting very detailed logs. We need to tell that user somehow that we have fallen back to using `info` logging and that error is a good way to communicate this.

vitorhugoro1 commented on level.go:
```diff
@@ -0,0 +1,87 @@
+	case LevelWarn:
+		return "warn"
+	case LevelError:
+		return "error"
```
Changed and added the LevelUndefined too

vitorhugoro1 commented on level.go:
```diff
@@ -0,0 +1,87 @@
+	LevelError Level = "error"
+)
+
+func (l Level) String() string {
```
Changed

vitorhugoro1 commented on level.go:
```diff
@@ -0,0 +1,87 @@
+	LevelInfo Level = "info"
+	// Log Level [warn].Print errors and logs which need attention on instrumentation.
+	LevelWarn Level = "warn"
+	// Log Level [error].Print errors generated by instrumentation.
```
Changed

vitorhugoro1 commented on level.go:
```diff
@@ -0,0 +1,87 @@
+)
+
+// Level defines the log level which instrumentation uses.
+type Level string
```
Makes sense, changed to LogLevel

vitorhugoro1 commented on CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@ OpenTelemetry Go Automatic Instrumentation adheres to [Semantic Versioning](http
 - Support Go `v1.21.10`. ([#824](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/824))
 - Support Go `v1.22.3`. ([#824](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/824))
 - Support `google.golang.org/grpc` `1.65.0-dev`. ([#827](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/827))
+- Add support to log level through command line flag. ([#842](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/842))
```
Moved to unreleased section and added the recommended

vitorhugoro1 commented:
Sorry for the late guys, applied the changes @MrAlias suggested and rebase with the main branch

MrAlias commented on level.go:
```diff
@@ -0,0 +1,94 @@
+type LogLevel string
+
+const (
+	// LevelUndefined is an unset log level, it should not be used.
+	LevelUndefined LogLevel = ""
```
Having this unexported will help prevent users from accidentally using it. It is only added to easily test unset or undefined logging levels.

```suggestion
	// levelUndefined is an unset log level, it should not be used.
	levelUndefined LogLevel = ""
```

vitorhugoro1 replied:
Changed

MrAlias commented on level.go:
```diff
@@ -0,0 +1,94 @@
+	"fmt"
+)
+
+// Level defines the log level which instrumentation uses.
```
```suggestion
// LogLevel defines the log level which instrumentation uses.
```

vitorhugoro1 replied:
Changed

MrAlias commented on level.go:
```diff
@@ -0,0 +1,94 @@
+const (
+	// LevelUndefined is an unset log level, it should not be used.
+	LevelUndefined LogLevel = ""
+	// LevelDebug sets the logging level to log all messages.
+	LevelDebug LogLevel = "debug"
+	// LevelInfo sets the logging level to log only informational, warning, and error messages.
+	LevelInfo LogLevel = "info"
+	// LevelWarn sets the logging level to log only warning and error messages.
+	LevelWarn LogLevel = "warn"
+	// LevelError sets the logging level to log only error messages.
+	LevelError LogLevel = "error"
```
```suggestion
	// LogLevelDebug sets the logging level to log all messages.
	LogLevelDebug LogLevel = "debug"
	// LogLevelInfo sets the logging level to log only informational, warning, and error messages.
	LogLevelInfo LogLevel = "info"
	// LogLevelWarn sets the logging level to log only warning and error messages.
	LogLevelWarn LogLevel = "warn"
	// LogLevelError sets the logging level to log only error messages.
	LogLevelError LogLevel = "error"
```

vitorhugoro1 replied:
Changed

MrAlias commented on level.go:
```diff
@@ -0,0 +1,94 @@
+	LevelError LogLevel = "error"
+)
+
+// String returns the string encoding of the Level l.
```
```suggestion
// String returns the string encoding of the LogLevel l.
```

vitorhugoro1 replied:
Changed

MrAlias commented on level.go:
```diff
@@ -0,0 +1,94 @@
+	return true
+}
+
+// ParseLevel return a new LogLevel using text, and will return err if inputted text is not accepted.
+func ParseLevel(text string) (LogLevel, error) {
```
```suggestion
// ParseLogLevel return a new LogLevel parsed from text. A non-nil error is returned if text is not a valid LogLevel.
func ParseLogLevel(text string) (LogLevel, error) {
```

vitorhugoro1 replied:
Changed

MrAlias commented on level.go:
```diff
@@ -0,0 +1,94 @@
+	}
+}
+
+// UnmarshalText applies the LogLevel type when inputted text is valid.
+func (l *LogLevel) UnmarshalText(text []byte) error {
+	if ok := l.unmarshalText(bytes.ToLower(text)); ok {
+		return nil
+	}
+
+	return l.validate()
+}
+
+func (l *LogLevel) validate() error {
+	if l == nil {
+		return errors.New("can't parse nil values")
+	}
+
+	if !l.unmarshalText(bytes.ToLower([]byte(l.String()))) {
+		return errors.New("log level value is not accepted")
+	}
+
+	return nil
+}
+
+func (l *LogLevel) unmarshalText(text []byte) bool {
+	switch string(text) {
+	case "debug":
+		*l = LevelDebug
+	case "info":
+		*l = LevelInfo
+	case "warn":
+		*l = LevelWarn
+	case "error":
+		*l = LevelError
+	default:
+		return false
+	}
+
+	return true
+}
```
This makes redundant calls to `unmarshalText`. Refactoring `validate` and `unmarshalText` will simplify:

```suggestion
// UnmarshalText applies the LogLevel type when inputted text is valid.
func (l *LogLevel) UnmarshalText(text []byte) error {
	*l = LogLevel(bytes.ToLower(text))
	return l.validate()
}

var errInvalidLogLevel = errors.New("invalid LogLevel")

func (l *LogLevel) validate() error {
	if l == nil {
		return errors.New("nil LogLevel")
	}

	switch *l {
	case LevelDebug, LevelInfo, LevelWarn, LevelError:
		// Valid.
	default:
		return fmt.Errorf("%w: %s", errInvalidLogLevel, l.String())
	}
	return nil
}
```

`TestValidate` can then be updated to just:

```go
func TestValidate(t *testing.T) {
	l := LogLevel("notexist")
	assert.ErrorIs(t, l.validate(), errInvalidLogLevel)
}
```

vitorhugoro1 replied:
Applied the suggestion and simplified TestValidate

MrAlias commented on instrumentation.go:
```diff
@@ -83,6 +94,10 @@ func newLogger() logr.Logger {
 	}
 
+	if logErr != nil {
+		logger.V(2).Info("error to parse log level, changed to LevelInfo by default")
```
```suggestion
		logger.Error("invalid log level; using LevelInfo instead", zap.Error(logErr), zap.String("input", logLevel.String()))
```

MrAlias commented on instrumentation.go:
```diff
@@ -383,6 +401,16 @@ func WithEnv() InstrumentationOption {
+			var e error
+			level, e := ParseLogLevel(l)
+
+			if err == nil {
```
```suggestion
			if e == nil {
```

vitorhugoro1 replied:
Nice, changed!

MrAlias commented on instrumentation.go:
```diff
@@ -490,3 +518,17 @@ func WithLoadedIndicator(indicator chan struct{}) InstrumentationOption {
 }
+
+// WithLogLevel returns an [InstrumentationOption] that will configure
+// an [Instrumentation] with the logger level visibility defined as inputed.
```
```suggestion
// an [Instrumentation] to use the provided logging level.
```

vitorhugoro1 replied:
Changed

MrAlias commented on level.go:
```diff
@@ -0,0 +1,79 @@
+type LogLevel string
+
+const (
+	// LevelUndefined is an unset log level, it should not be used.
```
```suggestion
	// logLevelUndefined is an unset log level, it should not be used.
```

vitorhugoro1 replied:
Changed

MrAlias commented on level_test.go:
Bump.

vitorhugoro1 replied:
Added tests to cover ParseLogLevel

MrAlias commented on level_test.go:
```diff
@@ -0,0 +1,56 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLevel(t *testing.T) {
```
Bump.

vitorhugoro1 replied:
Changed

RonFed commented:
@MrAlias can you please have another look?

vitorhugoro1 commented:
@MrAlias can you review again, please?
